extension/src/Template/TaskRunner.ts

```diff
@@ -0,0 +1,112 @@
+import * as vscode from "vscode";
+import * as path from "path";
+import * as fs from "fs";
+import * as VSCodeFunctions from "../VSCodeFunctions";
+import { TaskRunnerItem } from "./TaskRunnerItem";
+import * as SettingsLoader from "../Settings/SettingsLoader";
+
+export class TaskRunner {
+  private workspaceFilePath = "";
+  constructor(public taskList: TaskRunnerItem[]) {}
+
+  public get requiredTasks(): TaskRunnerItem[] {
+    return this.taskList.filter((t) => t.required === true);
+  }
+
+  async testRequired(): Promise<void> {
+    const missingCommands: string[] = [];
+    for (const task of this.requiredTasks) {
+      if (!(await VSCodeFunctions.commandExists(task.command))) {
+        missingCommands.push(`${task.description}: ${task.command}`);
+      }
+    }
+
+    if (missingCommands.length > 0) {
+      throw new Error(
+        `Could not find ${
+          missingCommands.length
+        } required command(s). ${missingCommands.join(", ")}`
+      );
+    }
+  }
+
+  async execute(task: TaskRunnerItem): Promise<void> {
+    await this.testRequired();
+    if (task.openFile) {
+      const openFilePath = path.join(this.workspaceFilePath, task.openFile);
+      if (!fs.existsSync(openFilePath)) {
+        throw new Error(
+          `Could not open file ${task.openFile} for command ${task.command}`
+        );
+      }
+      await vscode.workspace.openTextDocument(openFilePath);
+    }
+    await vscode.commands.executeCommand(task.command).then(
+      () => {
+        return;
```

Comment:
Since we're running arbitrary commands I'm not sure I can expect this to be executed. I haven't found anything to confirm this so I guess I'll try it out with the trusty old `reload window` :)

Reply:
Trusty reload window my ass :sweat_smile: We need to delete the file before running the command `onFullFilled` is not triggered if the window reloads. I guess we could try but I think we might risk an eternal loop and it would probably be very confusing to the user.

Reply:
Also an important note, avoid `forEach` when running async commands like this. False positives be everywhere :fearful:

Reply:
> Trusty reload window my ass :sweat_smile: We need to delete the file before running the command `onFullFilled` is not triggered if the window reloads. I guess we could try but I think we might risk an eternal loop and it would probably be very confusing to the user.

😅
But I still think we should delete file _after_ command is run by default. But we might need to create a list of hardcoded commands (as reload window and open workspace) that we delete file before running command. Also a setting in the al.template.json could let the user decide for other commands... The setting could be named 'restartsVSCode' or similar?

Reply:
> Also an important note, avoid `forEach` when running async commands like this. False positives be everywhere :fearful: 

Yes, that is probably why I moved away from the `forEach` construct earlier. It made me crazy...

Reply:
> > Trusty reload window my ass sweat_smile We need to delete the file before running the command `onFullFilled` is not triggered if the window reloads. I guess we could try but I think we might risk an eternal loop and it would probably be very confusing to the user.
> 
> sweat_smile But I still think we should delete file _after_ command is run by default. But we might need to create a list of hardcoded commands (as reload window and open workspace) that we delete file before running command. Also a setting in the al.template.json could let the user decide for other commands... The setting could be named 'restartsVSCode' or similar?

Yeah that's a good idea. I guess I burnt my self too bad to dare to think of such ingenuity :laughing:  I'll implement it :)